### ISSUE TITLE
Support JDK EA versions for toolchain downloads

### ISSUE THREAD

**Issue by bmuskalla** [gradle/gradle#14814]
Currently, EA versions are not supported as they require a different URL.

**Comment by Siedlerchr** [comment]
This would be a really good feature for testing upcoming releases

**Comment by Siedlerchr** [comment]
Still valid

**Comment by eskatos** [comment]
Thank you for your interest in Gradle!

This feature request is in the backlog of the relevant team, but this area of Gradle is currently not a focus one, so it might take a while before it gets implemented.

**Comment by koppor** [comment]
Background: The team of the Java JDK offers a page "Quality Outreach", where products (both closed and open source) can report whether their tool works with the latest early access build. This is important to support Java development. To enable an ease testing with EA versions, it would be great if gradle supported EA versions directly.

Deep link to the Quality Outreach page: https://wiki.openjdk.org/display/quality/Quality+Outreach.

Side note: @big-guy Maybe you could report them that gradle works with Java 21 GA.

**Comment by koppor** [comment]
A workaround is shown by the JUnit5 team: https://github.com/junit-team/junit5/blob/main/.github/workflows/cross-version.yml. Note that this requires also a [custom gradle plugin](https://github.com/junit-team/junit5/blob/d4adab8e08e9c73ecbfdef21638cfe3022c61fe5/gradle/plugins/common/src/main/kotlin/junitbuild.java-toolchain-conventions.gradle.kts) to enable passing the desired toolchain using `-PjavaToolchainVersion`.

---

Using ea versions is easy possible with the [Foojay Toolchains Plugin](https://github.com/gradle/foojay-toolchains?tab=readme-ov-file#foojay-toolchains-plugin), because Java 23 can be requested (https://github.com/foojayio/discoapi/issues/90). Just use `23` as tool chain version:

```groovy
java {
    sourceCompatibility = JavaVersion.VERSION_21
    targetCompatibility = JavaVersion.VERSION_21
    toolchain {
        languageVersion = JavaLanguageVersion.of(23)
    }
}
```

Note that you need to set `sourceCompatiblity` and `targetCompatibility` as otherwise somewhere within gradle there is an "unsupported class version" error. At least at my project

**Comment by ljacomet** [comment]
Given the evolution of the toolchain support in Gradle, this will never be a core feature.

Instead one could write a [dedicated provisioning plugin](https://docs.gradle.org/current/userguide/toolchain_plugins.html#toolchain_plugins) that gives access to EA Java versions.

**Comment by Sineaggi** [comment]
@ljacomet how would that work when the toolchain needs to be updated/replaced? Early access builds don't contain all code planned to be merged in, and it doesn't seem currently possible to pass additional toolchain request information around (such as `ea+21` from https://download.java.net/java/early_access/jdk25/21/GPL/openjdk-25-ea+21_linux-aarch64_bin.tar.gz or `valhalla+1-90` from https://download.java.net/java/early_access/valhalla/1/openjdk-23-valhalla+1-90_macos-aarch64_bin.tar.gz). While those still contain the root version, toolchains cannot provision minor or patch builds, much less build numbers or build branches.

Imagine this, you write a toolchain that downloads an early access build but then you want to upgrade to a later early access version, or have a second toolchain that can download production-ready toolchains (like the foojay resolver plugin). Without additional context, it seems impossible to provide this functionality witha. provisioning plugin.

**Comment by ljacomet** [comment]
Correct, there are a number of toolchains improvements that are required for this to work properly:
* #22495
* #18896

Right now, users need to be aware of those limitations and delete the previous version to get a new version provisioned.

**Comment by Sineaggi** [comment]
@ljacomet thanks for linking those two issues and verifying the workaround.

**Comment by koppor** [comment]
@Sineaggi Another interesting might be: https://github.com/gradle/gradle/issues/29508. This is risen by https://mise.jdx.dev/lang/java.html#gradle-toolchains-detection. `mise` is a JDK manager, which introduces another indirection, but works nicely to manage JDK installations. With [`mise prune`](https://mise.jdx.dev/cli/prune.html) old JDKs are removed.
